### DISCUSSION TITLE
Documentation for AutoReadFile

### DISCUSSION THREAD

**Discussion by kimjohn1** [SCIInstitute/SCIRun#2246] [General]
Working my way through an introduction to SCIRun 5, I see that there is no description of the new AutoReadFile in Modules/DataIO of SCIRun 5 documentation.  I saw the couple of issues, and the reference to a similar .NET instruction in the discussion, but couldn't get the gist of how to use it from that.
Do you have a user description elsewhere?

Kim

**Comment by dcwhite** [comment]
That's a pretty new module so it lacks documentation. The idea is you can give it any file, and the module will try to figure out what the file type is and what type of data is in it (field versus matrix, etc), rather than the user. So it's a bit quicker to use if you know what format your input files are in.

**Reply by kimjohn1** [reply]
Ok, so I have a data file that is a single cell from a 3D vector field, file extension .vtk.  My task is to visualize various aspects of this data.  I can do that with Paraview, and use that app to get more details about the data if needed.

I have set up AutoReadFile to read this data, but don't know where to go from there.  Is there a tutorial I could look at, that uses this module?  If not, can you suggest a set of modules I could use to get this data into a form that would allow me to show the individual vectors?

Kim

**Reply by jab0707** [reply]
If it reads in find as a .vtk file from autoReadFile then the next step would be to visualize the field. Fields are the yellow pipes in SCIRun. So try using the show field and view scene modules from here and let us know what happens after that. Pipe them together like int he picture and click the view button ont he view scene module. Should look something like this:
![image](https://user-images.githubusercontent.com/38257234/112540881-52392200-8d78-11eb-8083-caf7044c3e2b.png)

**Reply by kimjohn1** [reply]
When I Execute All with just the AutoReadFile module in the Network View Panel, and cube.vtk selected as the file to open, I get no error.  With ShowField also present, and connected to AutoReadFile, Execute All produces an error indicating that there was no input to ShowField and an error specific to AutoReadFile, indicating:
````
[SCIRun::Core::tag_error_message*] = No available reader could handle the file type
````
I have attached the full error message.

Kim

[AutoReadFile_error.txt](https://github.com/SCIInstitute/SCIRun/files/6208248/AutoReadFile_error.txt)

**Reply by kimjohn1** [reply]
Additional note:

I also tried ReadFile and using the VtktoTriSurfField interpreter when I select the file to be read.  That didn't work either, and produced an error:
````
cube.vtk is not compatible with this importer. The file should contain points and triangle cells.
````

I had a look at the cube.vtk file and it is quite simple.  There is a VTK header for version 4 and then eight 3-tuples (float) that are the 3D vectors for each of the 8 vertex points of a cube.  The data itself is ASCII, and is identified as such in the header.  I guess the VtktoTriSurfField module is looking for a surface rather than a 3D vector field.

Let me know if there is something else I can try using AutoReadFile.

Kim

**Reply by kimjohn1** [reply]
Since the data was such a small file, I manually deleted the header text, saved the eight 3-tuples as a .txt file and read that using AutoReadFile.  It worked ok.  I used a CreateLatVol module to make an array that could be rendered:
![text_data](https://user-images.githubusercontent.com/61957753/112587525-55a5cb00-8dc3-11eb-9aa4-e8e5db72c023.png)

CreateLatVol interprets the 3-tuple data as x,y,z coordinates, so the visualization is just a depiction of 8 points.  What module (or set of modules) can I use to assemble the 8 sets of 3-tuples as vectors, positioned at the corners of a 2X2X2 cube?

KIm

**Reply by dcwhite** [reply]
SetFieldData is what you want, @jessdtate @jab0707 can help you with the specifics.

**Reply by jessdtate** [reply]
Yeah, getting the order from what you already have may be tricky.  However, You can have two files, one for the vector locations, and the other for the vector values, read them in independently, with readfield and readmatrix respectively,   and use setfielddata to assign the vectors to the points.  

I think the original problem is that the vtk file reader doesn't support vector field data.
